File: Puzzle15VB/Puzzle.vb
Public Class Puzzle

    Public Const ColumnCount As Integer = 4
    Public Const RowCount As Integer = 4

    Public Const BitmapWidth As Integer = 400
    Public Const BitmapHeight As Integer = 400

    Public Const CellPixelWidth As Integer = BitmapWidth \ ColumnCount
    Public Const CellPixelHeight As Integer = BitmapHeight \ RowCount

    Private ReadOnly _backColor As Color = Color.Black
    Private _cells As Cell()
    Private _emptyCell As EmptyCell

    Public Event ImageUpdated(sender As Object, e As EventArgs)

    Public ReadOnly Property Img

    Public Sub New()
        Img = New Bitmap(BitmapWidth, BitmapHeight)
        _emptyCell = New EmptyCell(New Point(ColumnCount - 1, RowCount - 1))
        _cells = Enumerable.Range(0, ColumnCount * RowCount - 1).Select(Function(i) New Cell(i + 1, New Point(i Mod ColumnCount, i \ ColumnCount), $"img{i + 1}")).ToArray()
        UpdateVisual()
    End Sub

    Public Sub Shuffle()
        Me.Shuffle(Date.Now.Millisecond)
    End Sub

    Public Sub Shuffle(seed As Integer)
        Dim rand = New Random(seed)
        Dim nums = Enumerable.Range(0, _cells.Length).Select(Function(x) Tuple.Create(x, rand.Next())).OrderBy(Function(x) x.Item2).Select(Function(x, i) Tuple.Create(x.Item1, i))
        For Each num In nums
            _cells(num.Item1).Position = New Point(num.Item2 Mod ColumnCount, num.Item2 \ RowCount)
        Next
        _emptyCell.Position = New Point(ColumnCount - 1, RowCount - 1)
        If IsCompleted() Then
            Shuffle(seed + 1)
        End If
        UpdateVisual()
    End Sub

    Public Sub ToComplete()
        For i = 0 To _cells.Length - 1
            _cells(i).Position = New Point(i Mod ColumnCount, i \ ColumnCount)
        Next
        _emptyCell.Position = New Point(ColumnCount - 1, RowCount - 1)
        UpdateVisual()
    End Sub

    Public Sub MoveCell(targetCell As Point)
        Dim target = _cells.FirstOrDefault(Function(cell) cell.Position = targetCell)
        target.Position = _emptyCell.Position
        _emptyCell.Position = targetCell
        UpdateVisual()
    End Sub

    Public Function IsCompleted() As Boolean
        Return _cells.All(Function(cell) (cell.Number - 1) = (cell.PosY * ColumnCount + cell.PosX))
    End Function

    Public Function CanMove(targetCell As Point) As Boolean
        Dim target = _cells.FirstOrDefault(Function(cell) cell.Position = targetCell)
        If target Is Nothing Then
            Return False
        End If
        Return Math.Abs(_emptyCell.PosX - target.PosX) + Math.Abs(_emptyCell.PosY - target.PosY) = 1
    End Function


    Private Sub UpdateVisual()
        Using g As Graphics = Graphics.FromImage(Img)
            g.Clear(_backColor)
            For Each cell In _cells
                g.DrawImage(cell.Img, New Point(cell.PosX * CellPixelWidth, cell.PosY * CellPixelHeight))
            Next
        End Using
        RaiseEvent ImageUpdated(Me, EventArgs.Empty)
    End Sub
End Class

File: Puzzle15/Puzzle.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Puzzle15.Properties;

namespace Puzzle15
{
    public class Puzzle
    {
        public const int ColumnCount = 4;
        public const int RowCount = 4;

        public const int BitmapWidth = 400;
        public const int BitmapHeight = 400;

        public const int CellPixelWidth = BitmapWidth / ColumnCount;
        public const int CellPixelHeight = BitmapHeight / RowCount;

        private readonly Color _backColor = Color.Black;
        private readonly Cell[] _cells;
        private readonly EmptyCell _emptyCell;

        public event EventHandler ImageUpdated;

        public Bitmap Image { get; private set; }

        public Puzzle()
        {
            Image = new Bitmap(BitmapWidth, BitmapHeight);
            _emptyCell = new EmptyCell(new Point(ColumnCount - 1, RowCount - 1));
            _cells = Enumerable.Range(0, ColumnCount * RowCount - 1)
                              .Select(i => new Cell(i + 1, new Point(i % ColumnCount, i / ColumnCount), $"img{i + 1}"))
                              .ToArray();
            UpdateVisual();
        }

        public void Shuffle() => Shuffle((int)DateTime.Now.Ticks);

        public void ToComplete()
        {
            for (int i = 0; i < _cells.Length; i++)
            {
                _cells[i].Position = new Point(i % ColumnCount, i / ColumnCount);
            }
            _emptyCell.Position = new Point(ColumnCount - 1, RowCount - 1);
            UpdateVisual();
        }

        public void Shuffle(int seed)
        {
            var random = new Random(seed);
            const int swapCount = 50 * 2;


            var nums = Enumerable.Range(0, _cells.Length).ToArray();

            for (int i = 0; i < swapCount; i++)
            {
                var a = random.Next() % nums.Length;
                var b = random.Next() % (nums.Length - 1);
                if (b >= a) b++;

                (nums[a], nums[b]) = (nums[b], nums[a]);
            }



            foreach (var (x, i) in nums.Select((p, i) => (p, i)))
            {
                _cells[x].Position = new Point(i % ColumnCount, i / RowCount);
            }
            _emptyCell.Position = new Point(ColumnCount - 1, RowCount - 1);
            if (IsCompleted()) { Shuffle(seed + 1); }
            UpdateVisual();
        }

        /// <summary>セルを移動させます</summary>
        /// <param name="clickedPixel">クリックされたセル</param>
        public void MoveCell(Point targetCell)
        {
            var target = _cells.FirstOrDefault(cell => cell.Position == targetCell);
            target.Position = _emptyCell.Position;
            _emptyCell.Position = targetCell;
            UpdateVisual();
        }

        public bool IsCompleted()
        {
            return _cells.All(cell => cell.Number - 1 == cell.PosY * ColumnCount + cell.PosX);
        }

        public bool CanMove(Point targetCell)
        {
            var target = _cells.FirstOrDefault(cell => cell.Position == targetCell);
            if(target == null) { return false; }
            return Math.Abs(_emptyCell.PosX - target.PosX) + Math.Abs(_emptyCell.PosY - target.PosY) == 1;
        }

        private void UpdateVisual()
        {
            using (var g = Graphics.FromImage(Image))
            {
                g.Clear(_backColor);
                foreach (var cell in _cells)
                {
                    g.DrawImage(cell.Image, new Point(cell.PosX * CellPixelWidth, cell.PosY * CellPixelHeight));
                }
            }
            ImageUpdated?.Invoke(this, EventArgs.Empty);
        }
    }
}
